### PR TITLE
Dropout prob extremal patch

### LINKED ISSUE #1799: Dropout with prob == 0 doesn't validate consistently
### 🐛 Describe the bug

The following script doesn't validate consistently on TOT. It seems we may still be dropping out some values even though probability == 0. I think this may be because of: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/ops/composite.cpp#L31 which maybe should be `le` not `lt`?
```
import functools
import random
from typing import List

import torch
import torch.nn.functional as F

def composite_definition(
    input1: torch.Tensor,
    input2: torch.Tensor,
    weight: torch.Tensor,
    bias1: torch.Tensor,
    bias2: torch.Tensor,
    normalization_axis: int,
    dropout_prob: float,
) -> torch.Tensor:
    bias1_out = input1 + bias1
    dropout_out = F.dropout(bias1_out, 0.0, True)
    norm_input = dropout_out + input2
    norm_output = F.layer_norm(norm_input, (input1.size(normalization_axis),), weight, bias2)
    return norm_output

# Setup initial tensors and parameters
input_size = [64, 128, 1024]
device = "cuda"
dtype = torch.float32

# Create sample inputs
input1 = torch.randn(*input_size, device=device, dtype=dtype, requires_grad=True)
input2 = torch.rand_like(input1).requires_grad_()
 
# Precompute a grad output tensor, for this example it's the same size as the inputs
grad_output = torch.rand_like(input1)
 
# Randomly initialize the model parameters
weight = torch.nn.Parameter(torch.randn(input_size[2], dtype=dtype, device=device))
bias1 = torch.nn.Parameter(torch.randn(input_size[2], dtype=dtype, device=device))
bias2 = torch.nn.Parameter(torch.randn(input_size[2], dtype=dtype, device=device))

parameters = [input1, input2, weight, bias1, bias2]
ref_composite = composite_definition(input1, input2, weight, bias1, bias2, normalization_axis=2, dropout_prob=0.0)

scripted_composite_definition = torch.jit.script(composite_definition)

for i in range(20):
  scripted = scripted_composite_definition(input1, input2, weight, bias1, bias2, normalization_axis=2, dropout_prob=0.0)
  print("output abs max {}".format((ref_composite - scripted).abs().max()))
```

### Versions

TOT

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1799

Don't seem to impact perf in the example from the issue on a volta, but does look like it increases register usage.
Alternative is to slightly change the logic of mask generation for dropout, changing `lt` to `le`. But that means we'll diverge even further from aten RNG.

### CODE REVIEW

**Comment by jjsjann123** [comment]
We should add @IvanYashchuk to our project so we can request review from him. cc'ing @csarofeen

File: torch/csrc/jit/codegen/cuda/ops/composite.cpp [hunk 1]
```diff
@@ -29,6 +29,13 @@ ForwardDropoutResult dropout(TensorView* x, Val* prob, Val* scale) {
 
   auto rand_vals = randlike(x);
   auto mask = lt(rand_vals, prob);
+  // p == 0.0, set mask as False so everything is dropped.
```
Comment (IvanYashchuk) [inline torch/csrc/jit/codegen/cuda/ops/composite.cpp]: Is it needed? `lt(rand_vals, 0.0)` should always be all False because there shouldn't be any negative values generated by `randlike`.

File: torch/csrc/jit/codegen/cuda/ops/composite.cpp [hunk 2]
```diff
@@ -29,6 +29,13 @@ ForwardDropoutResult dropout(TensorView* x, Val* prob, Val* scale) {
 
   auto rand_vals = randlike(x);
```
Comment (IvanYashchuk) [inline torch/csrc/jit/codegen/cuda/ops/composite.cpp]: Alternatively, we could replace ones with zeros in `rand_vals` to change the interval from (0.0, 1.0] to [0.0, 1.0).
```suggestion
  auto rand_vals = randlike(x);
  auto rand_vals_new_interval = where(eq(rand_vals, IrBuilder::create<Double>(1.0)), IrBuilder::create<Double>(0.0), rand_vals);
```
Not sure, if using where or bitwise functions is better.

File: torch/csrc/jit/codegen/cuda/ops/composite.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ops/composite.cpp]: Good catch~
No it's not. Added it earlier for debug when I forgot we have `p` flipped :face_with_spiral_eyes:

File: torch/csrc/jit/codegen/cuda/ops/composite.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ops/composite.cpp]: `where` looks cleaner, I also think it should handle broadcast properly as well~ Let me try that~

Comment (IvanYashchuk) [reply torch/csrc/jit/codegen/cuda/ops/composite.cpp]: Using `where` inside `randlike` would also fix https://github.com/csarofeen/pytorch/issues/1807.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ops/composite.cpp]: That does sound right as well, are we pushing towards that fix?

Having no idea how RNG works. To my naive mind, it only messes with some corner distribution when we squash 1->0, so I'm somewhat comfortable that nobody would notice the difference of the distribution....

Comment (IvanYashchuk) [reply torch/csrc/jit/codegen/cuda/ops/composite.cpp]: I would push towards that fix.

It's difficult to notice this kind of difference, like the problem of dropout was there for some time already.
It's fine swapping 1.0 with 0.0 because mathematically every number has an equal probability to be generated and we won't mess up any mathematical properties.

CuPy does the same:
https://github.com/cupy/cupy/blob/18fab32f3d347fc13f86821cc72964515f6e5c4f/cupy/random/_generator.py#L619
```
    _mod1_kernel = _core.ElementwiseKernel(
        '', 'T x', 'x = (x == (T)1) ? 0 : x', 'cupy_random_x_mod_1')
```

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ops/composite.cpp]: Updated to that. I'm fine with the change. A trivial question doesn't clapping 1 to 0 means that we'll have twice as many 0s as other numbers? But that's just a minor thing so not really trying to dive into that rabbit hole.

Does the update look better?

**Comment by jjsjann123** [comment]
Looks like there's some failing tests after putting the where to rand_like. I'll let tests finish first and patch them afterwards.

**Comment by jjsjann123** [comment]
cleaned up some string format for float~ waiting on tests to finish.
Meanwhile, the PR is good for review~

File: torch/csrc/jit/codegen/cuda/ops/composite.cpp (hunk 2 above)
Comment (IvanYashchuk) [reply torch/csrc/jit/codegen/cuda/ops/composite.cpp]: There wouldn't be twice as many 0s, because there should be none generated by the Philox RNG (implemented here https://github.com/csarofeen/pytorch/blob/282c42902bff07f759cddbbe619249cf5e7c5281/torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu#L16), at least usually it's the case that 0 is excluded.

**Review by IvanYashchuk — APPROVED** [review]
Changes to `randlike` look good to me. Let's hope no test gets broken.

**Comment by jjsjann123** [comment]
Tests passed locally, patched some floating point prints. I'll merge once lintrunner finishes.
